graphql/batch_test.go

```diff
@@ -0,0 +1,368 @@
+				if set == nil {
+					return nil, errors.New("Expected to have selectionSet")
+				}
+				myMap := make(map[int]*Object, len(o))
+				for idx, val := range o {
+					val.Key = args.Prefix + val.Key
+					myMap[idx] = val
+				}
+				return myMap, nil
+			},
+			GiveValueFallbackFunc: func(ctx context.Context, o Object, args struct{ Prefix string }, set *graphql.SelectionSet) (*Object, error) {
+				if set == nil {
+					return nil, errors.New("Expected to have selectionSet")
+				}
+				o.Key = args.Prefix + o.Key
+				return &o, nil
+			},
+			GiveQuery: `
+			{
+				objects {
+					key
+					value(prefix: "test") {
+						key
+					}
+				}
+			}`,
+			WantResultJSON: `
+			{"objects": [
+			{"key": "key1", "value": {"key": "testkey1"}}
+			]}
+			`,
+		},
+		{
+			Name: "run with lots of objects",
+			GiveObjectFunc: func(ctx context.Context) []*Object {
+				return []*Object{
+					{Key: "key1"},
+					{Key: "key2"},
+					{Key: "key3"},
+					{Key: "key4"},
+					{Key: "key5"},
+					{Key: "key6"},
+					{Key: "key7"},
+					{Key: "key8"},
+				}
+			},
+			GiveValueFunc: func(ctx context.Context, o map[int]*Object, args struct{ Prefix string }, set *graphql.SelectionSet) (map[int]*Object, error) {
+				if set == nil {
+					return nil, errors.New("Expected to have selectionSet")
+				}
+				myMap := make(map[int]*Object, len(o))
+				for idx, val := range o {
+					val.Key = args.Prefix + val.Key
+					myMap[idx] = val
+				}
+				return myMap, nil
+			},
+			GiveValueFallbackFunc: func(ctx context.Context, o Object, args struct{ Prefix string }, set *graphql.SelectionSet) (*Object, error) {
+				if set == nil {
+					return nil, errors.New("Expected to have selectionSet")
+				}
+				o.Key = args.Prefix + o.Key
+				return &o, nil
+			},
+			GiveQuery: `
+			{
+				objects {
+					key
+					value(prefix: "test") {
+						key
+					}
+				}
+			}`,
+			WantResultJSON: `
+			{"objects": [
+			{"key": "key1", "value": {"key": "testkey1"}},
+			{"key": "key2", "value": {"key": "testkey2"}},
+			{"key": "key3", "value": {"key": "testkey3"}},
+			{"key": "key4", "value": {"key": "testkey4"}},
+			{"key": "key5", "value": {"key": "testkey5"}},
+			{"key": "key6", "value": {"key": "testkey6"}},
+			{"key": "key7", "value": {"key": "testkey7"}},
+			{"key": "key8", "value": {"key": "testkey8"}}
+			]}
+			`,
+		},
+	}
+
+	const (
+		OldExecutor           = "OldExecutor"
+		NewExecutorNoBatching = "NewExecutorNoBatching"
+		NewExecutorBatching   = "NewExecutorBatching"
+	)
+	conditions := []string{OldExecutor, NewExecutorNoBatching, NewExecutorBatching}
+	for _, tt := range tests {
+		for _, cond := range conditions {
+			t.Run(fmt.Sprintf("%s %s", tt.Name, cond), func(t *testing.T) {
+				builder := schemabuilder.NewSchema()
+				builder.Query().FieldFunc("objects", tt.GiveObjectFunc)
+
+				obj := builder.Object("object", Object{})
+				obj.BatchFieldFunc("value", tt.GiveValueFunc, tt.GiveValueFallbackFunc, func(ctx context.Context) bool {
```

Comment:
could we make the batch field func API a little more forward facing? how about we ask that the fallback-resolver is registered through .FieldFunc, and the decision function is passed to .BatchFieldFunc with a special option schemabuilder.UseBatchFieldFuncFunction(...)?

alternatively, and this might be a little nutty, how about we make it the job of the application code to build multiple schemas, and with and one without batching?

or, we let folks register both, and we create a top-level function on the schema builder to supply the batch decision function which will take a type and field name?

Reply:
i like separate register calls .FieldFunc, .BatchFieldFunc and a global decision func that takes type and field name. that feels the most user-friendly for porting to batch and for deprecating fallbacks